### PR TITLE
Lab3 Finished

### PR DESCRIPTION
**Pull request by shenganzhang**

### CODE REVIEW

File: game.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+bool Game::shiftTiles(Direction dir) {
+    return grid.shiftTiles(dir);
+}
+
+bool Game::checkWin() const {
+    bool flag = false;
+    for (int vValue = 0; vValue < grid.getHeight(); ++vValue) {
+        for (int hValue = 0; hValue < grid.getWidth(); ++hValue) {
+            if (!grid.isEmpty({vValue,hValue}) && (grid.getSquare({vValue,hValue})->points == WINPOINTS)){
```
Comment (TimothyGe) [inline game.cpp]: `> WINPOINTS` can also be a winning situation.

File: grid.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                    if (shiftTile(checkDst(dir, Point{i, j}, *this), Point{i, j})) {
+                        flag = true;
+                    }
+                }
+            }
+        }
+            break;
+
+        case DOWN:
+        case RIGHT: {
```
Comment (TimothyGe) [inline grid.cpp]: Good catch on the difference of iteration order.

**Comment by shenganzhang** [comment]
I have updated other files for lab3 and it works well on my own laptop. If it is still not OK, perhaps I misunderstood some of the requirements of the lab or there are some imperceptible bugs.

**Comment by zianke** [comment]
Hi, I've replaced my `Grid::collapseTiles()` function with your implementation. It seems that you are doing some kind of "recursive collapsing".

For example, suppose the initial state is
```
|       4	|       2	|       8	|        	|        	|        	|

|      16	|       2	|        	|       2	|        	|        	|

|       2	|      32	|       8	|        	|        	|        	|

|       2	|        	|        	|        	|        	|        	|

|       4	|        	|        	|        	|        	|        	|
```
Then after an `UP` operation, your next state would be
```
|       4	|       4	|      16	|       2	|        	|        	|

|      16	|      32	|        	|        	|        	|        	|

|       8	|        	|        	|        	|        	|        	|

|        	|        	|        	|        	|        	|        	|

|        	|        	|        	|        	|        	|        	|
```
but the correct one should have two Tile "4"s in the first column.
```
|       4	|       4	|      16	|       2	|        	|        	|

|      16	|      32	|        	|        	|        	|        	|

|       4	|        	|        	|        	|        	|        	|

|       4	|        	|        	|        	|        	|        	|

|        	|        	|        	|        	|        	|        	|
```
In other words, any tile can be collapsed at most once in a single move.

**Comment by zianke** [comment]
Thanks for your pull request. Most of your implementations are correct.
While there might be more elegant implementations of other functions, your `Grid::upgradeTile()` looks perfect to me, so I've merged it into the main repository.

**Pull request merged**